fe/fe-core/src/main/java/org/apache/doris/ldap/LdapManager.java

```diff
@@ -0,0 +1,197 @@
+
+    private final LdapClient ldapClient = new LdapClient();
+
+    private final Map<String, LdapUserInfo> ldapUserInfoCache = Maps.newHashMap();
+
+    private final ReentrantReadWriteLock lock = new ReentrantReadWriteLock();
+
+    private long lastTimeStamp = System.currentTimeMillis();
+
+    // If the user exists in LDAP, the LDAP information of the user is returned; otherwise, null is returned.
+    public LdapUserInfo getUserInfo(String fullName) {
+        if (!checkParam(fullName)) {
+            return null;
+        }
+        LdapUserInfo ldapUserInfo = getUserInfoFromCache(fullName);
+        if (ldapUserInfo != null && !ldapUserInfo.checkTimeout()) {
+            return ldapUserInfo;
+        }
+        return updateUserInfo(fullName);
+    }
+
+    public boolean doesUserExist(String fullName) {
+        if (!checkParam(fullName)) {
+            return false;
+        }
+        return !Objects.isNull(getUserInfo(fullName));
+    }
+
+    public boolean checkUserPasswd(String fullName, String passwd) {
+        String userName = ClusterNamespace.getNameFromFullName(fullName);
+        if (!LdapConfig.ldap_authentication_enabled || Strings.isNullOrEmpty(userName) || Objects.isNull(passwd)) {
+            return false;
+        }
+        LdapUserInfo ldapUserInfo = getUserInfo(fullName);
+        if (Objects.isNull(ldapUserInfo)) {
+            return false;
+        }
+
+        if (ldapUserInfo.isSetPasswd() && ldapUserInfo.getPasswd().equals(passwd)) {
+            return true;
+        }
+        boolean isRightPasswd = ldapClient.checkPassword(userName, passwd);
+        if (!isRightPasswd) {
+            return false;
+        }
+        updatePasswd(ldapUserInfo, passwd);
+        return true;
+    }
+
+    public boolean checkUserPasswd(String fullName, String passwd, String remoteIp, List<UserIdentity> currentUser) {
+        if (checkUserPasswd(fullName, passwd)) {
+            currentUser.add(UserIdentity.createAnalyzedUserIdentWithIp(fullName, remoteIp));
+            return true;
+        }
+        return false;
+    }
+
+    private boolean checkParam(String fullName) {
+        return LdapConfig.ldap_authentication_enabled && !Strings.isNullOrEmpty(fullName) && !fullName.equalsIgnoreCase(
+                PaloAuth.ROOT_USER) && !fullName.equalsIgnoreCase(PaloAuth.ADMIN_USER);
+    }
+
+    private LdapUserInfo updateUserInfo(String fulName) {
+        String cluster = ClusterNamespace.getClusterNameFromFullName(fulName);
+        String userName = ClusterNamespace.getNameFromFullName(fulName);
+        if (Strings.isNullOrEmpty(userName)) {
+            return null;
+        } else if (!ldapClient.doesUserExist(userName)) {
+            removeUserIfExist(fulName);
+            return null;
+        }
+        checkTimeoutCleanCache();
+
+        LdapUserInfo ldapUserInfo = new LdapUserInfo(fulName, false, "", getLdapGroupsPrivs(userName, cluster));
+        lock.writeLock().lock();
```

Comment:
Would it be better to use writeLock() and writeUnlock()?

Reply:
good idea